src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,354 @@
+            INodeStats nodeStats,
+            IAsyncLoopFactory asyncLoopFactory)
+        {
+            Guard.NotNull(chain, nameof(chain));
+            Guard.NotNull(loggerFactory, nameof(loggerFactory));
+            Guard.NotNull(provenBlockHeaderRepository, nameof(provenBlockHeaderRepository));
+            Guard.NotNull(nodeLifetime, nameof(nodeLifetime));
+            Guard.NotNull(nodeStats, nameof(nodeStats));
+            Guard.NotNull(asyncLoopFactory, nameof(asyncLoopFactory));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.chain = chain;
+            this.provenBlockHeaderRepository = provenBlockHeaderRepository;
+            this.nodeLifetime = nodeLifetime;
+
+            this.lockObject = new object();
+            this.PendingBatch = new Dictionary<int, ProvenBlockHeader>();
+            this.Cache = new MemorySizeCache<int, ProvenBlockHeader>(this.MemoryCacheSizeLimitInBytes);
+
+            this.asyncLoopFactory = asyncLoopFactory;
+
+            this.performanceCounter = new BackendPerformanceCounter(dateTimeProvider);
+            nodeStats.RegisterStats(this.AddBenchStats, StatsType.Benchmark, 100);
+        }
+
+        /// <inheritdoc />
+        public async Task InitializeAsync()
+        {
+            await this.provenBlockHeaderRepository.InitializeAsync().ConfigureAwait(false);
+
+            this.TipHashHeight = this.provenBlockHeaderRepository.TipHashHeight;
+
+            this.logger.LogDebug("Initialized ProvenBlockHeader block tip at '{0}'.", this.TipHashHeight);
+
+            this.asyncLoop = this.asyncLoopFactory.Run("ProvenBlockHeaders job", async token =>
+            {
+                // Save pending items.
+                await this.SaveAsync().ConfigureAwait(false);
+
+                this.logger.LogTrace("(-)[IN_ASYNC_LOOP]");
+            },
+            this.nodeLifetime.ApplicationStopping,
+            repeatEvery: TimeSpan.FromMinutes(1),
+            startAfter: TimeSpan.FromMinutes(1));
+        }
+
+        /// <inheritdoc />
+        public async Task<ProvenBlockHeader> GetAsync(int blockHeight)
+        {
+            ProvenBlockHeader header = null;
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                if (!this.Cache.TryGetValue(blockHeight, out header))
+                {
+                    // Check the repository.
+                    header = await this.provenBlockHeaderRepository.GetAsync(blockHeight).ConfigureAwait(false);
+
+                    this.Cache.AddOrUpdate(blockHeight, header, header.HeaderSize);
+                }
+            }
+
+            return header;
+        }
+
+        /// <inheritdoc />
+        public async Task<List<ProvenBlockHeader>> GetAsync(int fromBlockHeight, int toBlockHeight)
+        {
+            Guard.Assert(toBlockHeight >= fromBlockHeight);
+
+            var headersInCache = new Dictionary<int, ProvenBlockHeader>();
```

Comment:
I suggest you dont need headersInCache, just use provenHeadersOutput.

Reply:
Ok, done.